### ISSUE TITLE
Phase 2: Neural Network Forward Pass on Vulkan GPU

### ISSUE THREAD

**Issue by ulises-jeremias** [vlang/vtl#59]
## Motivation

Phase 1 provides the compute primitives: GEMM, element-wise ops, broadcast on Vulkan. Phase 2 wires these into the **full neural network forward pass** on GPU — Conv2D, Linear, LSTM, Attention, Pool, LayerNorm — replacing all 7-deep nested loops and CPU BLAS calls with GPU kernels.

This is the phase that transforms VTL from a CPU-only library into a **GPU-accelerated tensor library** with real end-to-end performance gains for inference and training.

### Files to create

1. **`compute/softmax.v`** — `softmax_vulkan[T](input, dim)` — full softmax on GPU
2. **`compute/layernorm.v`** — `layernorm_vulkan[T](input, gamma, beta, eps)` — LayerNorm on GPU
3. **`compute/reduction.v`** — `sum_vulkan`, `mean_vulkan`, `max_vulkan` along axis on GPU
4. **`compute/conv2d.v`** — `conv2d_vulkan_forward` — im2col pre-pass + GEMM on Vulkan
5. **`nn/internal/activations_vulkan.v`** — fused forward activations (ReLU, Sigmoid, Tanh, GELU, LeakyReLU, ELU, Swish, Mish)
6. **`nn/layers/conv2d.v`** — `Conv2DLayer.forward()` with `$if defined(vulkan)` dispatch (Im2col + GEMM)
7. **`nn/layers/lstm.v`** — `LSTMLayer.forward()` with Vulkan GEMM/gates (if time permits)
8. **`nn/layers/attention.v`** — `MultiHeadAttentionLayer.forward()` with scaled-dot-product on Vulkan
9. **`nn/layers/pool.v`** — `AveragePool2D`, `GlobalAvgPool2D` on Vulkan

### Files to modify

- **`nn/layers/linear.v`** — Already wired in Phase 1; verify and extend `backward()` CPU fallback
- **`nn/layers/layer_norm.v`** — Add Vulkan path for `LayerNormLayer.forward()`
- **`nn/layers/embedding.v`** — Lookup on GPU (scatter/gather kernel)
- **`nn/layers/pos_encoding.v`** — Sin/cos on GPU

### API Contracts

```v
// Conv2D forward on Vulkan
pub fn conv2d_vulkan_forward[T](input &VulkanTensor[T], weight &VulkanTensor[T], ...)!&VulkanTensor[T]

// LayerNorm on Vulkan
pub fn layernorm_vulkan[T](x &VulkanTensor[T], gamma &VulkanTensor[T], beta &VulkanTensor[T], eps f64) !&VulkanTensor[T]

// Softmax on Vulkan
pub fn softmax_vulkan[T](input &VulkanTensor[T], dim int) !&VulkanTensor[T]

// Pooling on Vulkan
pub fn avgpool2d_vulkan[T](input &VulkanTensor[T], kernel []int, ...) !&VulkanTensor[T]
pub fn global_avgpool2d_vulkan[T](input &VulkanTensor[T]) !&VulkanTensor[T]
```

### Vulkan Kernels to Add

1. **Softmax kernel** (row-wise: exp - max - sum - normalize)
2. **LayerNorm kernel** (compute mean, variance, normalize, scale+bias)
3. **Im2col kernel** (NHWC -> Im2col matrix for GEMM-based conv)
4. **Reduction kernels** (sum, mean, max along axis)
5. **Pooling kernels** (avg pool, max pool — optional, can use reduction)
6. **Embedding lookup** (gather/scatter)

### Benchmark Targets

- **Conv2D**: >= 5x speedup vs VTL CPU Conv2D (7-nested-loop baseline)
- **Linear**: >= 20x speedup vs VSL BLAS matmul
- **LSTM**: >= 10x speedup vs pure-V loop
- **Attention**: >= 10x speedup vs CPU attention (if implemented)
- **LayerNorm**: >= 5x speedup vs CPU `.map()` loop

### Conv2D Strategy (Im2col + GEMM)

```
Input [N, C, H, W]
  |
  v
Im2col transform -> [N, out_h, out_w, C*k_h*k_w]   (GPU kernel)
  |
  v
Weight [K, C, k_h, k_w] -> [K, C*k_h*k_w]         (GPU reshape)
  |
  v
GEMM ([N*out_h*out_w] x [C*k_h*k_w]) @ ([K] x [C*k_h*k_w]) -> [N*out_h*out_w, K]
  |
  v
Reshape -> [N, out_h, out_w, K] -> [N, K, out_h, out_w]
  |
  v
Add bias (element-wise kernel)
```

### Dependencies

- Phase 1 `compute/gemm.v`, `compute/gemv.v`, `compute/elementwise.v` must be complete
- `storage/vulkan.v`, `vulkan_tensor.v` must be stable

## Checklist

- [ ] `compute/softmax.v`: softmax_vulkan[T]
- [ ] `compute/layernorm.v`: layernorm_vulkan[T]
- [ ] `compute/reduction.v`: sum_vulkan, mean_vulkan, max_vulkan along axis
- [ ] `compute/conv2d.v`: im2col_vulkan + conv2d via GEMM
- [ ] `nn/internal/activations_vulkan.v`: all activations on Vulkan
- [ ] `nn/layers/conv2d.v`: forward() on Vulkan (Im2col + GEMM)
- [ ] `nn/layers/layer_norm.v`: forward() on Vulkan
- [ ] `nn/layers/attention.v`: forward() on Vulkan (scaled-dot-product)
- [ ] `nn/layers/pool.v`: avgpool2d, global_avgpool2d on Vulkan
- [ ] `nn/layers/embedding.v`: lookup on Vulkan (gather kernel)
- [ ] `nn/layers/pos_encoding.v`: sin/cos on Vulkan
- [ ] Benchmarks: Conv2D, Linear, LSTM, LayerNorm speedups vs CPU
- [ ] Integration test: end-to-end model forward pass on Vulkan

Parent: #57
Labels: enhancement, gpu, phase-2
